Drivers/Src/CANbus.c

```diff
@@ -1,61 +1,197 @@
+        OSSemPend(
+            &CANMail_Sem4,
+            0,
+            OS_OPT_PEND_NON_BLOCKING,
+            &timestamp,
+            &err
+        );
+    }
+    if (err != OS_ERR_NONE){
+        return ERROR;
+    }
+
+
+    uint8_t txdata[8];
+    uint8_t datalen;
+
+    switch(id){
+    //Handle 64bit precision case (no idx)
+       case MC_BUS:
+       case VELOCITY:
+       case MC_PHASE_CURRENT:
+       case VOLTAGE_VEC:
+       case CURRENT_VEC:
+       case BACKEMF:
+       case TEMPERATURE:
+       case ODOMETER_AMPHOURS:
+            datalen = 8;
+            memcpy(&txdata,&payload.data.d,sizeof(payload.data.d));
+
+
+    //Handle 8bit precision case (0b0000xxxx) (no idx)
+       case CAR_STATE:
+            datalen = 1;
+            memcpy(&txdata,&payload.data.b,sizeof(payload.data.b)); //copy into txdata array  the 8 bits of the payload
+
+    }
+
+    OSMutexPend( //ensure that tx line is available
+        &CANbus_TxMutex,
+        0,
+        OS_OPT_PEND_BLOCKING,
+        &timestamp,
+        &err
+    );
+    if (err != OS_ERR_NONE){
+        //couldn't lock tx line
+        return ERROR;
+    }
+    //tx line locked
+
+    int retval = BSP_CAN_Write(CAN_1, id, txdata, datalen);
+
+    OSMutexPost( //unlock the TX line
+        &CANbus_TxMutex,
+        OS_OPT_POST_NONE,
+        &err
+    );
+
+    if (retval){
+        return SUCCESS;
+    } else {
+        return ERROR;
+    }
+}
+
+
+/**
+ * @brief   Checks if the CAN ID matches with expected ID and then copies message to given buffer array
+ * @param   CAN line bus (variable buses not implemented, this function currently works with CAN1)
+ * @param   ID pointer to where to store the CAN id of the recieved msg
+ * @param   pointer to buffer array to store message
+ * @param   blocking whether or not this read should be blocking
+ * @return  1 if ID matches and 0 if it doesn't
+ */
+
+ErrorStatus CANbus_Read(uint8_t* buffer,CAN_blocking_t blocking){
+    CPU_TS timestamp;
+    OS_ERR err;
+
+    if(blocking == CAN_BLOCKING){
+        OSSemPend( //check if the queue actually has anything
+            &CANBus_RecieveSem4,
+            0,
+            OS_OPT_PEND_BLOCKING,
+            &timestamp,
+            &err
+        );
+    } else {
+        OSSemPend(
+            &CANBus_RecieveSem4,
+            0,
+            OS_OPT_PEND_NON_BLOCKING,
+            &timestamp,
+            &err
+        );
+    }
+    if(err != OS_ERR_NONE) {
+        return ERROR;
+    }
+
+    OSMutexPend( //ensure that RX line is available
+        &CANbus_RxMutex,
+        0,
+        OS_OPT_PEND_BLOCKING,
+        &timestamp,
+        &err
+    );
+    if (err != OS_ERR_NONE){ 
+        //couldn't lock RX line
+        return ERROR;
+    }
+
+    //GETMSG
+    CANId_t ret;
+    uint8_t status = BSP_CAN_Read(CAN_1,&ret,buffer);
+    
+
+    OSMutexPost( //unlock RX line
+        &CANbus_RxMutex,
+        OS_OPT_POST_1,
+        &err
+    );
+    if(status&&(ret == CHARGE_ENABLE)){
```

Comment:
I think that this is the wrong layer to be performing this check. Should probably either be doing this in the Apps layer or through the hardware filters in the BSP layer, or even better: in both (in case the system needs to pay attention to more messages in the future).